WOR-430

blakery opened the pull request:
[WOR-430](https://broadworkbench.atlassian.net/browse/WOR-430)

Emit metrics for big query spend reports.


- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

blakery committed:
simplified unit tests to require less mocking

blakery committed:
made spend report aggregation static and remove slickDatasource from some tests

makes it obvious that the aggregation doesn't use any class resources, and shaves multiple seconds off unit tests

blakery committed:
finished removing usages of TestDriverComponent

Tests for spend reports are now completely mocked, with no external db calls

blakery committed:
add metrics to spend reporting service

blakery committed:
moved query specific logic to query building function

blakery committed:
move query setup to a separate function

blakery committed:
add convenience methods to create rawls exceptions

avoids the intermediary step of having to explicitly create the error report

blakery committed:
fiddling with names types and formatting

blakery committed:
update workbench lib versions to pick up new functionality

blakery committed:
get full big query job info and report metrics from job stats

blakery committed:
clean up error codes

blakery committed:
minor cleanup missed in prev commit

blakery committed:
output cache stats as a ratio

marctalbott commented on core/src/main/scala/org/broadinstitute/dsde/rawls/spendreporting/SpendReportingService.scala:
```diff
@@ -66,28 +75,28 @@ object SpendReportingService {
-      .map { case (googleProjectId, projectRows) =>
-        val workspaceName = workspaceProjectsToNames.getOrElse(
-          googleProjectId,
+      .map { case (projectId, projectRows) =>
```
`projectId` can lead to confusion between Terra billing projects and Google projects

blakery replied:
Before refactors, I was finding it difficult to read because of sheer volume of text floating around.

I just needed to take a second pass at the refactor to clean it up.

marctalbott commented on core/src/main/scala/org/broadinstitute/dsde/rawls/spendreporting/SpendReportingService.scala:
```diff
@@ -55,6 +63,7 @@ object SpendReportingService {
       .setScale(currency.getDefaultFractionDigits, RoundingMode.HALF_EVEN)
       .toString()
 
+    type Key = SpendReportingAggregationKey
```
This feels like unnecessary misdirection to me -- if I want to know what a `Key` is, I have to click through this first to get to the definition of `SpendReportingAggregationKey`

marctalbott commented on core/src/main/scala/org/broadinstitute/dsde/rawls/spendreporting/SpendReportingService.scala:
```diff
@@ -30,19 +32,25 @@ object SpendReportingService {
+  val BigQueryKey = "bigQuery"
+  val BigQueryCacheMetric = "cache"
+  val BigQueryBytesProcessedMetric = "processed"
+  type BillingToWorkspaceNames = Map[String, WorkspaceName]
```
Not sure about this name. `BillingToWorkspaceNames` implies Billing projects to me which is not what this is. I'd prefer to remove the type alias entirely and revert back to `Map[GoogleProjectId, WorkspaceName]` for the sake of type clarity or `Map[String, WorkspaceName]` at the least

marctalbott reviewed (commented):
The logic of it all looks good, but I have reservations regarding some of the variable and type name changes. I personally prefer longer variable and type names if they make it easier to read and understand the code -- even if they're more annoying to deal with when writing. Happy to discuss if you'd like

blakery committed:
second pass at aggregation refactor

this time I found a way to make the functionality of different aggregation keys generic
